src/pal/pal_haiku.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#if defined(__HAIKU__)
+#  include "pal_posix.h"
+
+#  include <sys/mman.h>
+
+namespace snmalloc
+{
+  /**
+   * Platform abstraction layer for Haiku.  This provides features for this
+   * system.
+   */
+  class PALHaiku : public PALPOSIX<PALHaiku>
+  {
+  public:
+    /**
+     * Bitmap of PalFeatures flags indicating the optional features that this
+     * PAL supports.
+     *
+     */
+    static constexpr uint64_t pal_features = PALPOSIX::pal_features;
+
+    /**
+     * Notify platform that we will not be needing these pages.
+     * Haiku does not provide madvise call per say only the posix equivalent.
+     */
+    void notify_not_using(void* p, size_t size) noexcept
+    {
+      SNMALLOC_ASSERT(is_aligned_block<page_size>(p, size));
+      posix_madvise(p, size, POSIX_MADV_DONTNEED);
+    }
+
+    /**
+     *  OS specific function for zeroing memory
+     *  using MAP_NORESERVE for explicit over commit appliance.
+     */
+    template<bool page_aligned = false>
+    void zero(void* p, size_t size)
+    {
+      if (page_aligned || is_aligned_block<page_size>(p, size))
+      {
+        SNMALLOC_ASSERT(is_aligned_block<page_size>(p, size));
+        void* r = mmap(
+          p,
+          size,
+          PROT_READ | PROT_WRITE,
+          MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED | MAP_NORESERVE,
+          -1,
+          0);
+
+        if (r != MAP_FAILED)
+          return;
+      }
+
+      bzero(p, size);
+    }
+
+    /**
+     * Reserve memory using MAP_NORESERVE for explicit
+     * over commit applicance.
+     */
+    std::pair<void*, size_t> reserve_at_least(size_t size)
```

Comment:
As far as I can tell, the only difference between this and the POSIX PAL is the addition of the `MAP_NORESERVE` flag (the same is true for the `zero` function).  Almost the same is true for the Apple PAL: it differs only in the file descriptor used for anonymous memory mappings.

We should probably allow the template passed to the POSIX PAL to provide extra default `mmap` flags and specify the `mmap` file descriptor, so that we're not copying and pasting code as much.  Not essential for this PR, but if you don't do it then please file an issue to track it.

Reply:
I ll do a PR then.